### PR TITLE
journalctl: in --follow mode watch stdout for POLLHUP/POLLERR and exit

### PR DESCRIPTION
**Pull request by poettering**
Fixes: #9374

### CODE REVIEW

File: src/journal/journalctl.c [hunk 1]
```diff
@@ -2037,14 +2037,46 @@ static int sync_journal(void) {
...
+
+        r = sd_journal_get_timeout(j, &timeout);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine time to wait for: %m");
+
+        if (ppoll(pollfds, ELEMENTSOF(pollfds), timeout == USEC_INFINITY ? NULL : timespec_store(&ts, timeout), NULL) < 0)
+                return log_error_errno(errno, "Couldn't wait for journal event: %m");
+
+        if (pollfds[1].revents & (POLLHUP|POLLERR)) { /* STDOUT has been closed? */
+                log_error("Standard output has been closed. Cancelling.");
```
Comment (keszybz) [inline src/journal/journalctl.c]: I think "Cancelling." should be removed. It's an internal detail, and the user of the program will have no idea what will be "cancelled.".

This should be a `log_debug` message. It's a routine occurrence and not an error.

File: src/journal/journalctl.c [hunk 2]
```diff
@@ -2037,14 +2037,46 @@ static int sync_journal(void) {
...
+
+        assert(j);
+        assert(poll_fd >= 0);
+
+        /* Much like sd_journal_wait() but also keeps an eye on STDOUT, and exits as soon as we see a POLLHUP on that,
+         * i.e. when it is closed. */
+
+        r = sd_journal_get_timeout(j, &timeout);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine time to wait for: %m");
```
Comment (keszybz) [inline src/journal/journalctl.c]: Maybe "Failed to determine journal waiting time: %m" ?

**Comment by poettering** [comment]
force pushed a new version. no change besides the two log msg rewordings and downgrading of the log level. taking liberty to set the green label hence.

**Comment by keszybz** [comment]
I tested this quite a bit in the previous version and it seems all OK.

**Pull request merged**